pkg/services/ngalert/state/historian/core_test.go

```diff
@@ -97,6 +98,82 @@ func TestShouldRecord(t *testing.T) {
+	}
+	transition := func(from eval.State, fromReason string, to eval.State, toReason string) state.StateTransition {
+		return state.StateTransition{
+			PreviousState:       from,
+			PreviousStateReason: fromReason,
+			State:               &state.State{State: to, StateReason: toReason},
+		}
+	}
+
+	t.Run("transitions between Normal and Normal(NoData) not recorded if NoData->OK", func(t *testing.T) {
+		r := rule()
+		r.NoDataState = models.OK
+		forward := transition(eval.Normal, "", eval.Normal, models.StateReasonNoData)
+		backward := transition(eval.Normal, models.StateReasonNoData, eval.Normal, "")
+
+		require.False(t, shouldRecordAnnotation(r, forward), "Normal -> Normal(NoData) should be false")
+		require.False(t, shouldRecordAnnotation(r, backward), "Normal(NoData) -> Normal should be false")
+	})
+
+	t.Run("transitions between Normal and Normal(NoData) are recorded if NoData not mapped to OK", func(t *testing.T) {
+		noDataToAlerting := rule()
+		noDataToAlerting.NoDataState = models.Alerting
+		noDataToNoData := rule()
+		noDataToNoData.NoDataState = models.NoData
+		forward := transition(eval.Normal, "", eval.Normal, models.StateReasonNoData)
+		backward := transition(eval.Normal, models.StateReasonNoData, eval.Normal, "")
+
+		require.True(t, shouldRecordAnnotation(noDataToAlerting, forward), "Normal -> Normal(NoData) should be true")
+		require.True(t, shouldRecordAnnotation(noDataToAlerting, backward), "Normal(NoData) -> Normal should be true")
+		require.True(t, shouldRecordAnnotation(noDataToNoData, forward), "Normal -> Normal(NoData) should be true")
+		require.True(t, shouldRecordAnnotation(noDataToNoData, backward), "Normal(NoData) -> Normal should be true")
+	})
+
+	t.Run("other normal transitions involving NoData still recorded even if NoData->OK", func(t *testing.T) {
+		r := rule()
+		r.NoDataState = models.OK
+		pauseForward := transition(eval.Normal, models.StateReasonNoData, eval.Normal, models.StateReasonPaused)
+		pauseBackward := transition(eval.Normal, models.StateReasonPaused, eval.Normal, models.StateReasonNoData)
+		errorForward := transition(eval.Normal, models.StateReasonNoData, eval.Normal, models.StateReasonError)
+		errorBackward := transition(eval.Normal, models.StateReasonError, eval.Normal, models.StateReasonNoData)
+		missingSeriesBackward := transition(eval.Normal, models.StateReasonMissingSeries, eval.Normal, models.StateReasonNoData)
+
+		require.True(t, shouldRecordAnnotation(r, pauseForward), "Normal(NoData) -> Normal(Paused) should be true")
+		require.True(t, shouldRecordAnnotation(r, pauseBackward), "Normal(Paused) -> Normal(NoData) should be true")
+		require.True(t, shouldRecordAnnotation(r, errorForward), "Normal(NoData) -> Normal(Error) should be true")
+		require.True(t, shouldRecordAnnotation(r, errorBackward), "Normal(Error) -> Normal(NoData) should be true")
+		require.True(t, shouldRecordAnnotation(r, missingSeriesBackward), "Normal(MissingSeries) -> Normal(NoData) should be true")
```

Comment:
I think we can drop all transitions from to Error\NoData and keep MissingSeries, Paused and Updated